pkg/dataobj/index/builder.go

```diff
@@ -230,35 +247,46 @@ func (p *Builder) processRecord(record *kgo.Record) {
 		level.Error(p.logger).Log("msg", "failed to unmarshal metastore event", "err", err)
 		return
 	}
-	p.bufferedEvents[event.Tenant] = append(p.bufferedEvents[event.Tenant], *event)
-	level.Info(p.logger).Log("msg", "buffered new event for tenant", "count", len(p.bufferedEvents[event.Tenant]), "tenant", event.Tenant)
-
-	if len(p.bufferedEvents[event.Tenant]) >= p.cfg.EventsPerIndex {
-		if !slices.Contains(p.mCfg.Storage.EnabledTenantIDs, event.Tenant) {
-			// TODO(benclive): Remove this check once builders handle multi-tenancy when building indexes.
-			level.Info(p.logger).Log("msg", "skipping index build for disabled tenant", "tenant", event.Tenant)
-			p.bufferedEvents[event.Tenant] = p.bufferedEvents[event.Tenant][:0]
-			return
-		}
-		err := p.buildIndex(p.bufferedEvents[event.Tenant][:len(p.bufferedEvents[event.Tenant])])
+	p.bufferedEvents[record.Partition] = append(p.bufferedEvents[record.Partition], *event)
+	level.Info(p.logger).Log("msg", "buffered new event for partition", "count", len(p.bufferedEvents[record.Partition]), "partition", record.Partition)
+
+	if len(p.bufferedEvents[record.Partition]) >= p.cfg.EventsPerIndex {
+		// Create a sub-context for the index build, so we can cancel it if the partition is revoked.
+		var calculationCtx context.Context
+		p.activeCalculationPartition = record.Partition
```

Comment:
As above, we need mutex around this to avoid data race between this and the on partition revoked callback.